Permissions-Policy-Report-Only and Reporting API updates (#43853)
mdn/content: files/en-us/web/api/reporting_api/index.md

- Occurrence of crashes.
- Occurrence of user-agent interventions (when the browser blocks something your code is trying to do because it is deemed a security risk for example, or just plain annoying, like auto-playing audio).

The purpose of the Reporting API is to provide a consistent reporting mechanism that can be used to make such information available to developers in the form of reports represented by JavaScript objects.

> Reports can be obtained within a page using JavaScript via reporting observers, or can be sent to a remote server endpoint.

Reviewer: ```suggestion
Reports can be retrieved via JavaScript reporting observers or sent to a remote server endpoint.
```

The types of reports and the two reporting approaches are detailed in the sections below.

### Report types

Reports sent to reporting observers are instances of dictionary objects, such as {{domxref("COEPViolationReport")}}, {{domxref("DeprecationReport")}}, {{domxref("IntegrityViolationReport")}}, {{domxref("InterventionReport")}}, {{domxref("CSPViolationReport")}}, and {{domxref("PermissionsPolicyViolationReport")}}.
These all have an origin `url`, a `type`, and a `body` that is specific to the report type.
The type of report can be determined from its `type` property, which for the reports above would be `coep`, `deprecation`, `integrity-violation`, `intervention`, `csp-violation`, and `permissions-policy-violation`.

Reports sent to reporting endpoints and reporting observers are essentially the same.
The only difference is that server reports are JSON serializations of the objects that have additional `user_agent` and `age` fields.

> A list of documented report types and their corresponding report dictionary are given in the [`options.types`](/en-US/docs/Web/API/ReportingObserver/ReportingObserver#types) parameter passed to the `ReportingObserver()` constructor.

Reviewer: ```suggestion
A list of documented report types and their corresponding report dictionaries is given in the [`options.types`](/en-US/docs/Web/API/ReportingObserver/ReportingObserver#types) parameter passed to the `ReportingObserver()` constructor.
```

### Reporting observers

Reports can be obtained via {{domxref("ReportingObserver")}} objects created via JavaScript inside the website you are aiming to get reports on.

> This method is not as failsafe as sending reports to the server because any page crash could stop you retrieving the reports — but it is easier to set up, and more flexible.

Reviewer: ```suggestion
This method is not as failsafe as sending reports to the server because any page crash could stop you from retrieving the reports; it is, however, easier to set up, and more flexible.
```

| `coop`                         | `COOPViolationReport`                           | {{httpheader("Cross-Origin-Opener-Policy")}} (COOP) violations                             |
| `crash`                        | {{domxref("CrashReport")}}                      | Browser crash reports                                                                      |
| `csp-violation`                | {{domxref("CSPViolationReport")}}               | [Content Security Policy (CSP)](/en-US/docs/Web/HTTP/Guides/CSP) violations                |
| `deprecation`                  | {{domxref("DeprecationReport")}}                | Deprecated features used by the site.                                                      |
| `integrity-violation`          | {{domxref("IntegrityViolationReport")}}         | {{httpheader("Integrity-Policy")}} violations                                              |
| `intervention`                 | {{domxref("InterventionReport")}}               | Features blocked by the user agent, such as ads that significantly impact page performance |

> | `permissions-policy-violation` | {{domxref("PermissionsPolicyViolationReport")}} | {{httpheader("Permissions-Policy")}} violations                                            |

Author: @chrisdavidmills Thanks for the review. I accepted all your suggestions bar one, which has a comment.

The only other change is I re-wrote this section on report types to have a clear list.

Note that this is the _documented_ types. There are some I haven't tested yet, such as document policy and NetworkErrorReport. Those might require further updates - i.e. I have a feeling that NetworkErrorReport can't be sent to server endpoints so the text might need minor amendment. 

I think this is good to go. Updating `-Report-Only` headers for other HTTP headers will be a separate task done right at the end.

Reviewer: Yup, looks good to me.

### Reporting server endpoints

Reports can also be sent to remote _server endpoints_ by the user agent in a `POST` operation with a {{HTTPHeader("Content-Type")}} of `application/reports+json`.
They are serializations of the corresponding dictionary for each [report type](#report_types).
For example, CSP violation reports are a serialization of a {{domxref("CSPViolationReport")}} object.

Reports sent to endpoints can be retrieved independently of the running of the websites they relate to, which is useful — a crash for example could bring down a website and stop anything running, but a report could still be obtained to give the developer some clues as to why it happened.

> [!NOTE]
> There is no absolute guarantee of report delivery — a report could still fail to be collected if a serious error occurs.

The {{httpheader("Reporting-Endpoints")}} HTTP header is used to specify the name and URL for different endpoints that a user-agent has available to it for delivering reports.

> The endpoints can then be used on particular HTTP response headers to indicate the specific endpoint (or in some cases endpoints) that will be used for the associated report.

Reviewer: ```suggestion
The endpoints can then be specified within particular HTTP response headers to indicate the endpoint (or in some cases, endpoints) that associated reports will be delivered to.
```

Report types that don't have an associated HTTP header, such as `crash`, `deprecation`, and `intervention` reports, will usually send reports to the [`"default"` reporting endpoint](/en-US/docs/Web/HTTP/Reference/Headers/Reporting-Endpoints#default_reporting_endpoint) (this is just an endpoint named "default" specified using the `Reporting-Endpoints` header).

> The mechanism to specify server endpoint for each report is listed below.

Reviewer: ```suggestion
The mechanism to specify server endpoints for each report is listed below.
```

> - `coep`

Reviewer: The content seems fine, but the DL formatting is broken.

Author: That's because it isn't supposed to be a DL - a DL is problematic because there are multiple bullets in some cases and it is messy to write prefix text in some cases and not others. Fixed up as DL-like in https://github.com/mdn/content/pull/43853/changes/d07f69464991084a4b871f797fab2bc182eb0ce7

Reviewer: Right, Ok. This looks better now.
